README.md
# CoSaMP

CoSaMP algorithm in Haskell. 

[Documentation](https://augustunderground.github.io/cosamp)

## Use in Stack Project

Add this to `stack.yaml`:

```yaml
extra-deps:
 - github: augustunderground/cosamp
   commit: 264f4ec2f6a74d10a7cc484e0bbbe7f24d85d3cc
```

## Example

```haskell
import CoSaMP
import Numeric.LinearAlgebra
import Graphics.Plot

main :: IO ()
main      = mplot [t, s] >> mplot [tWin, xWin, xWin']
  where 
    n     = 4096 :: Int
    t     = linspace n (0.0, 1.0 :: Double)
    x     = sin ( 91 * 2 * pi * t ) + sin ( 412 * 2 * pi * t )
    p     = 128 :: Int
    aquis = cmap floor 
          $ scale (realToFrac n - 1) (randomVector 666 Uniform p) :: Vector I
    y     = flatten $ asRow x ?? (All, Pos aquis)
    φ     = dctOrtho (ident n :: Matrix Double)
    cφ    = φ ?? (Pos aquis, All)
    s     = cosamp cφ y 10 100 1.0e-10
    x'    = idctOrtho s
    wIdx  = idxs $ find (\i -> i >= 0.4 && i <= 0.5) t
    tWin  = flatten $ asRow t ?? (All, Pos wIdx)
    xWin  = flatten $ asRow x ?? (All, Pos wIdx)
    xWin' = flatten $ asRow x' ?? (All, Pos wIdx)
```

![](docs/result.png)

![](docs/spectrum.png)

## Build From Source

```
$ git clone https://github.com/augustunderground/cosamp
$ cd cosamp
$ stack build
$ stack install
$ stack run
```

# References and Resources

This is a Haskell implementation of the Compressive Sampling Matched Pursuit
(CoSaMP) algorithm as presented in Needell and Tropp's 2008
[paper](http://arxiv.org/abs/0803.2392).

There are also reference implementations available in:
- [MATLAB](https://www.mathworks.com/matlabcentral/fileexchange/32402-cosamp-and-omp-for-sparse-recovery).
- [Python](https://github.com/rfmiotto/CoSaMP) which is based on 


CHANGELOG.md
# Changelog for `cosamp`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


cosamp.py
import numpy as np
import scipy
import matplotlib.pyplot as plt

# def cosamp(Phi, u, s, tol=1e-10, max_iter=1000):
def cosamp(Phi, u, s, tol=1e-10, max_iter=10):
    max_iter -= 1 # Correct the while loop
    num_precision = 1e-12
    a = np.zeros(Phi.shape[1])
    v = u
    iter = 0
    halt = False
    while not halt:
        iter += 1
        y = abs(np.dot(np.transpose(Phi), v))
        Omega = [i for (i, val) in enumerate(y) if val > np.sort(y)[::-1][2*s] and val > num_precision] # equivalent to below
        #Omega = np.argwhere(y >= np.sort(y)[::-1][2*s] and y > num_precision)
        T = np.union1d(Omega, a.nonzero()[0])
        #T = np.union1d(Omega, T)
        b = np.dot( np.linalg.pinv(Phi[:,T]), u )
        iGood = (abs(b) > np.sort(abs(b))[::-1][s]) & (abs(b) > num_precision)
        T = T[iGood]
        a[T] = b[iGood]
        v = u - np.dot(Phi[:,T], b[iGood])
        halt = np.linalg.norm(v)/np.linalg.norm(u) < tol or \
               iter > max_iter
        print(f"Iteration {iter}: a = {a.nonzero()[0]}\n")
    return a

n = 4096 # number of measurements
t = np.linspace(0.0, 1.0, num=n)

x = np.sin(91*2*np.pi*t) + np.sin(412*2*np.pi*t) # original signal (to be reconstructed)

# randomly sample signal
p = 128 # random sampling (Note that this is one eighth of the Shannon–Nyquist rate!)
aquis = np.round((n-1) * np.random.rand(p)).astype(int)
y = x[aquis] # our compressed measurement from the random sampling

# Here {y} = [C]{x} = [C][Phi]{s}, where Phi is the inverse discrete cosine transform
Phi = scipy.fft.dct(np.eye(n), axis=0, norm='ortho')
CPhi = Phi[aquis,:]

# l1 minimization (through linear programming)
s = cosamp(CPhi, y, 10) # obtain the sparse vector through CoSaMP algorithm
xrec = scipy.fft.idct(s, axis=0, norm='ortho') # Reconstructed signal

s[s.nonzero()[0]]


ss = np.array([6.936376032725912, 27.474683924431055, -27.578700285137383, -11.337040416677038, -7.6673838856754575, -4.563676616381389, -4.8639213347650365, -6.087959111673908, 6.127047371932055, 10.446992690237028, 33.76464111472499, -14.967453149814236, -16.20380157256696, -4.713715149349071, 5.0523166106586945 ])



scipy.fft.idct(ss, axis=0, norm='ortho')

figw, figh = 7.0, 5.0 # figure width and height
plt.figure(figsize=(figw, figh))
plt.plot(t, s)
plt.title('Sparse vector $s$')
plt.show()

# Visualize the compressed-sensing reconstruction signal
figw, figh = 7.0, 5.0 # figure width and height
plt.figure(figsize=(figw, figh))
plt.plot(t, x,    'b', label='Original signal')
plt.plot(t, xrec, 'r', label='Reconstructed signal')
plt.xlim(0.4, 0.5)
legend = plt.legend(loc='upper center', shadow=True, fontsize='x-large')
# Put a nicer background color on the legend.
legend.get_frame().set_facecolor('C0')
plt.show()


src/CoSaMP.hs
{-# OPTIONS_GHC -Wall #-}

{-# LANGUAGE RecordWildCards #-}

-- | Compressive Sampling Matched Pursuit (CoSaMP): Iterative signal recovery
-- from incomplete and inaccurate samples" by Deanna Needell & Joel Tropp
module CoSaMP ( cosamp
              , dctOrtho
              , idctOrtho
              ) where

import Numeric.LinearAlgebra
import Numeric.LinearAlgebra.Devel        (ti)
import Prelude                     hiding ((<>), reverse)
import Data.List                          (nub)
import Control.Monad.State
import Statistics.Transform               (dct, idct)

-- | CoSaMP Algorithm State
data CoSaMPState = CoSaMPState 
    { p :: Double        -- ^ Precision
    , i :: Int           -- ^ Iteration
    , e :: Double        -- ^ Tolerance
    , φ :: Matrix Double -- ^ Sampling Matrix
    , u :: Vector Double -- ^ Noisy sampling vector
    , a :: Vector Double -- ^ s-sparse approximation of target signal @ iteration @i@
    , v :: Vector Double -- ^ Updated sample vector
    , s :: Int           -- ^ Sparsity level
    } deriving (Show)

-- | Compressive Sampling Matched Pursuit (CoSaMP)
cosamp :: Matrix Double -- ^ sampling matrix Φ
       -> Vector Double -- ^ noisy sample vector u
       -> Int           -- ^ sparsity level s
       -> Int           -- ^ Number of Iterations (default 1000)
       -> Double        -- ^ Tolerance (default 1.0e-10)
       -> Vector Double -- ^ s-sparse approsimation of the target signal
cosamp φ' u' s' i' e' = evalState (cosamp' False) st
  where
    p' = 1.0e-12
    a' = konst (0 :: Double) . snd $ size φ'
    st = CoSaMPState { p = p'
                     , i = i'
                     , e = e'
                     , φ = φ'
                     , u = u'
                     , a = a'
                     , v = u'
                     , s = s' }

-- | CoSaMP Algorithm in the CoSaMPState Monad
cosamp' :: Bool -> State CoSaMPState (Vector Double)
cosamp' True  = gets a
cosamp' False = do
    st@CoSaMPState{..} <- get
    let y  = abs $ tr φ #> v
        c  = sortVector y ! (size y - 1 - 2*s)
        ω  = find (\val -> val > c && val > p) y
        t  = nub $ ω ++ find (/= 0) a :: [IndexOf Vector]
        b  = pinv (φ ?? (All, Pos (idxs t))) #> u
        g  = sortVector (abs b) ! (size b - 1 - s)
        j  = idxs . find (\b'' -> b'' > g && b'' > p) $ abs b
        t' = flatten $ asRow (idxs t) ?? (All, Pos j)
        b' = flatten $ asRow b ?? (All, Pos j)
        a' = accum a const $ zip (map ti $ toList t') (toList b')
        v' = u - (φ ?? (All, Pos t') #> b')
        i' = i - 1
        h  = (norm_2 v / norm_2 u) < e || i <= 0
    put $ st { i = i', v = v', a = a' }
    cosamp' h

-- | DCT Type II where corresponding matrix coefficients are made orthonormal
dctOrtho :: Matrix Double -> Matrix Double
dctOrtho x = y 
  where
    n  = fst $ size x
    n' = realToFrac n :: Double
    t  = fromColumns . map dct $ toColumns x
    r0 = scale (1.0 / ( 2.0 * sqrt n')) (t ?? (Pos $ idxs [0], All))
    rs = scale (sqrt (2.0 / n') / 2.0) (t ?? (Pos $ idxs [1 .. (n - 1)], All))
    y  = r0 === rs

-- | Inverse DCT Type II where corresponding matrix coefficients are made orthonormal
idctOrtho :: Vector Double -> Vector Double
idctOrtho y = x
  where
    n = realToFrac $ size y :: Double
    x = scale (0.5 * sqrt (2.0 / n)) (idct y) 


app/Main.hs
{-# OPTIONS_GHC -Wall #-}

module Main (main) where

import CoSaMP
import Numeric.LinearAlgebra
import Graphics.Plot

main :: IO ()
main      = mplot [t, s] >> mplot [tWin, xWin, xWin']
  where 
    n     = 4096 :: Int
    t     = linspace n (0.0, 1.0 :: Double)
    x     = sin ( 91 * 2 * pi * t ) + sin ( 412 * 2 * pi * t )
    p     = 128 :: Int
    aquis = cmap floor 
          $ scale (realToFrac n - 1) (randomVector 666 Uniform p) :: Vector I
    y     = flatten $ asRow x ?? (All, Pos aquis)
    φ     = dctOrtho (ident n :: Matrix Double)
    cφ    = φ ?? (Pos aquis, All)
    s     = cosamp cφ y 10 100 1.0e-10
    x'    = idctOrtho s
    wIdx  = idxs $ find (\i -> i >= 0.4 && i <= 0.5) t
    tWin  = flatten $ asRow t ?? (All, Pos wIdx)
    xWin  = flatten $ asRow x ?? (All, Pos wIdx)
    xWin' = flatten $ asRow x' ?? (All, Pos wIdx)
